bin/pytorch_inference/Main.cc

```diff
@@ -90,12 +80,33 @@ void writeTensor(const torch::TensorAccessor<T, 1UL>& accessor,
     jsonWriter.EndArray();
 }
 
-template<typename T>
-void writeTensor(const torch::TensorAccessor<T, 2UL>& accessor,
+template<typename T, std::size_t N_DIMS>
+void writeTensor(const torch::TensorAccessor<T, N_DIMS>& accessor,
                  ml::core::CRapidJsonLineWriter<rapidjson::OStreamWrapper>& jsonWriter) {
+    jsonWriter.StartArray();
     for (int i = 0; i < accessor.size(0); ++i) {
         writeTensor(accessor[i], jsonWriter);
     }
+    jsonWriter.EndArray();
+}
+
+template<typename T>
+void writeInferenceResults(const torch::TensorAccessor<T, 3UL>& accessor,
+                           ml::core::CRapidJsonLineWriter<rapidjson::OStreamWrapper>& jsonWriter) {
+
+    jsonWriter.Key(INFERENCE);
+    writeTensor(accessor, jsonWriter);
+}
+
+template<typename T>
+void writeInferenceResults(const torch::TensorAccessor<T, 2UL>& accessor,
+                           ml::core::CRapidJsonLineWriter<rapidjson::OStreamWrapper>& jsonWriter) {
```

Comment:
Nice tidy up this is much clearer and also better to keep tensor writing generic IMO. (I was just wrong and one can of course overload functions whose arguments are class partial specialisations so no need to have the class bit either.) One thing I find interesting is you still have
```
template<typename T>
void writeTensor(const torch::TensorAccessor<T, 1UL>& accessor,
                 ml::core::CRapidJsonLineWriter<rapidjson::OStreamWrapper>& jsonWriter) {
    jsonWriter.StartArray();
    for (int i = 0; i < accessor.size(0); ++i) {
        jsonWriter.Double(static_cast<double>(accessor[i]));
    }
    jsonWriter.EndArray();
}
```
when you know `torch::TensorAccessor<T, 1UL>` only contains a single value. Is this intentional? It seems surprising to me.

Reply:
The 2nd template parameter `1UL` is number of dimensions, it may well be a scalar but it could also be a vector. Note `operator[]` returns an accessor with one fewer dimensions `torch::TensorAccessor<T, N -1>` or when `N == 1` the underlying data.

I really like this pattern matching aspect of C++ where there is a base case for a single dimension and the overloaded method handles the N > 1 cases. This would have taken much more code to write in Java.

Reply:
Oh sorry, my mistake, looking at the implementation again I realise I'd just misread the code.
> I really like this pattern matching aspect of C++

Yes and variadic templates make this sort of idiom even more powerful.